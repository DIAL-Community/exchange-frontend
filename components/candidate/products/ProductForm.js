import React, { useCallback, useContext, useEffect, useRef, useState } from 'react'
import { useIntl } from 'react-intl'
import { useRouter } from 'next/router'
import Link from 'next/link'
import { useMutation } from '@apollo/client'
import { FaSpinner } from 'react-icons/fa'
import ReCAPTCHA from 'react-google-recaptcha'
import { Controller, useForm } from 'react-hook-form'
import { validate } from 'email-validator'
import { ToastContext } from '../../../lib/ToastContext'
import { CREATE_CANDIDATE_PRODUCT } from '../../../mutations/product'
import Input from '../../shared/Input'
import ValidationError from '../../shared/ValidationError'
import { HtmlEditor } from '../../shared/HtmlEditor'
import { Unauthorized } from '../../shared/FetchStatus'
import UrlInput from '../../shared/UrlInput'
import { BREADCRUMB_SEPARATOR } from '../../shared/breadcrumb'
import { useUser } from '../../../lib/hooks'
import Select from '../../shared/Select'
import { getProductLicenseType } from '../../../lib/utilities'

const ProductForm = ({ candidateProduct }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const router = useRouter()

  const { user, isAdminUser } = useUser()

  const { showToast } = useContext(ToastContext)

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const captchaRef = useRef(null)

  const slug = candidateProduct?.slug ?? ''

  const licenseTypeOptions = getProductLicenseType(format)
  const [ossLicenseType, commercialLicenseType] = licenseTypeOptions

  const [createCandidateProduct, { reset }] = useMutation(CREATE_CANDIDATE_PRODUCT, {
    onError: () => {
      setMutating(false)
      showToast(format('candidate-product.submit.failure'), 'error', 'top-center')
      reset()
    },
    onCompleted: (data) => {
      const { createCandidateProduct: response } = data
      if (response?.errors?.length === 0) {
        setMutating(false)
        showToast(
          format('candidate-product.submit.success'),
          'success',
          'top-center',
          1000,
          null,
          () => {
            const nextPath = !isAdminUser ? '/products' : '/candidate/products/'
            router.push(nextPath)
          }
        )
      } else {
        setMutating(false)
        showToast(format('candidate-product.submit.failure'), 'error', 'top-center')
        reset()
      }
    }
  })

  const { handleSubmit, register, control, setValue, formState: { errors } } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: candidateProduct?.name,
      description: candidateProduct?.description,
      repository: candidateProduct?.repository || '',
      website: candidateProduct?.website || '',
      submitterEmail: candidateProduct?.submitterEmail,
      commercialProduct: candidateProduct?.commercialProduct ? commercialLicenseType : ossLicenseType,
      captcha: null
    }
  })

  const doUpsert = async (data) => {
    if (user) {
      setMutating(true)

      const captcha = captchaRef.current.getValue()

      const { userEmail, userToken } = user
      const { name, description, submitterEmail, website, repository, commercialProduct } = data

      const variables = {
        slug,
        name,
        website,
        repository,
        description,
        submitterEmail,
        commercialProduct: commercialProduct.value === 'commercial',
        captcha
      }

      createCandidateProduct({
        variables,
        context: {
          headers: {
            'Accept-Language': router.locale,
            Authorization: `${userEmail} ${userToken}`
          }
        }
      })
    }
  }

  const cancelForm = () => {
    setReverting(true)
    const nextPath = !isAdminUser ? '/products' : '/candidate/products/'
    router.push(nextPath)
  }

  // Building breadcrumbs
  const parentBreadcrumb = () =>
    isAdminUser
      ? (
        <Link href='/candidate/products'>
          <a className='text-dial-blue'>
            {format('candidateProduct.label')}
          </a>
        </Link>
      )
      : (
        <Link href='/products'>
          <a className='text-dial-blue'>
            {format('product.header')}
          </a>
        </Link>
      )

  const childBreadcrumb = (candidateProduct) =>
    !candidateProduct || !isAdminUser
      ? (
        <span className='text-dial-gray-dark'>
          {format('app.create')}
        </span>
      )
      : (
        <>
          <Link href={`/candidate/products/${candidateProduct.slug}`}>
            <a className='text-dial-blue'>
              {candidateProduct?.name}
            </a>
          </Link>
          {BREADCRUMB_SEPARATOR}
          <span className='text-dial-gray-dark'>
            {format('app.edit')}
          </span>
        </>
      )

  const generateBreadcrumb = (candidateProduct) => (
    <div className='bg-white pb-3 lg:py-4 whitespace-nowrap text-ellipsis overflow-hidden'>
      <Link href='/'>
        <a className='inline text-dial-blue h5'>
          {format('app.home')}
        </a>
      </Link>
      <div className='inline h5'>
        {BREADCRUMB_SEPARATOR}
        {parentBreadcrumb()}
        {BREADCRUMB_SEPARATOR}
        {childBreadcrumb(candidateProduct)}
      </div>
    </div>
  )

  useEffect(() => {
    register('captcha', { required: format('validation.required') })
  }, [register, format])

  return user ? (
    <div className='flex flex-col'>
      <div className='hidden lg:block px-8'>
        { generateBreadcrumb(candidateProduct) }
      </div>
      <div className='py-8 px-8'>
        <div id='content' className='sm:px-0 max-w-full mx-auto'>
          <form onSubmit={handleSubmit(doUpsert)}>
            <div className='bg-edit shadow-md rounded px-8 pt-6 pb-12 mb-4 flex flex-col gap-3'>
              <div className='text-2xl font-semibold text-dial-sapphire pb-4'>
                {format('candidateProduct.label')}
              </div>
              <div className='flex flex-col lg:flex-row gap-4'>
                <div className='w-full lg:w-1/2 flex flex-col gap-y-3'>
                  <div className='form-field-wrapper' data-testid='candidate-product-name'>
                    <label className='form-field-label required-field' htmlFor='name'>
                      {format('candidateProduct.name')}
                    </label>
                    <Input
                      {...register('name', { required: format('validation.required') })}
                      id='name'
                      placeholder={format('candidateProduct.name.placeholder')}
                      isInvalid={errors.name}
                    />
                    {errors.name && <ValidationError value={errors.name?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='candidate-product-website'>
                    <label className='form-field-label required-field' htmlFor='website'>
                      {format('candidateProduct.website')}
                    </label>
                    <Controller
                      name='website'
                      control={control}
                      rules={{ required: format('validation.required') }}
                      render={({ field: { value, onChange } }) => (
                        <UrlInput
                          value={value}
                          onChange={onChange}
                          id='website'
                          placeholder={format('candidateProduct.website.placeholder')}
                        />
                      )}
                    />
                    {errors.website && <ValidationError value={errors.website?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='candidate-product-repository'>
                    <label className='form-field-label required-field' htmlFor='repository'>
                      {format('candidateProduct.repository')}
                    </label>
                    <Controller
                      name='repository'
                      control={control}
                      rules={{ required: format('validation.required') }}
                      render={({ field: { value, onChange } }) => (
                        <UrlInput
                          value={value}
                          onChange={onChange}
                          id='repository'
                          placeholder={format('candidateProduct.repository.placeholder')}
                        />
                      )}
                    />
                    {errors.repository && <ValidationError value={errors.repository?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='candidate-product-email'>
                    <label className='form-field-label required-field' htmlFor='submitterEmail'>
                      {format('candidateProduct.email')}
                    </label>
                    <Input
                      type='email'
                      {...register('submitterEmail', {
                        required: format('validation.required'),
                        validate: value => validate(value) || format('validation.email')
                      })}
                      id='submitterEmail'
                      placeholder={format('candidateProduct.email.placeholder')}
                      isInvalid={errors.email}
                    />
                    {errors.email && <ValidationError value={errors.email?.message} />}
                  </div>
                  <div className='form-field-wrapper' data-testid='building-block-maturity'>
                    <label className='flex gap-x-2 mb-2 items-center self-start text-dial-sapphire'>
                      {format('product.license')}
                    </label>
                    <Controller
                      name='commercialProduct'
                      control={control}
                      render={({ field }) => (
                        <Select
                          {...field}
                          options={licenseTypeOptions}
                          placeholder={format('product.commercialProduct')}
                        />
                      )}
                    />
                  </div>
                  <ReCAPTCHA
                    sitekey={process.env.NEXT_PUBLIC_CAPTCHA_SITE_KEY}
                    ref={captchaRef}
                    onChange={value => {
                      setValue('captcha', value, { shouldValidate: true })
                    }}
                  />
                  {errors.captcha && <ValidationError value={errors.captcha?.message} />}
                </div>
                <div className='w-full lg:w-2/3' style={{ minHeight: '20rem' }} data-testid='candidate-product-description'>
                  <label className='block text-dial-sapphire flex flex-col gap-y-2'>
                    <p className='required-field'> {format('candidateProduct.description')}</p>
                    <Controller
                      name='description'
                      control={control}
                      rules={{ required: format('validation.required') }}
                      render={({ field: { value, onChange, onBlur } }) => {
                        return (
                          <HtmlEditor
                            editorId={`${name}-editor`}
                            onBlur={onBlur}
                            onChange={onChange}
                            initialContent={value}
                            isInvalid={errors.description}
                            placeholder={format('candidateProduct.description.placeholder')}
                          />
                        )
                      }}
                    />
                    {errors.description && <ValidationError value={errors.description?.message} />}
                  </label>
                </div>
              </div>
              <div className='flex flex-wrap font-semibold text-xl lg:mt-8 gap-3'>
                <button
                  type='submit'
                  data-testid='submit-button'
                  className='submit-button'
                  disabled={mutating || reverting}
                >
                  {format('candidateProduct.submit')}
                  {mutating && <FaSpinner className='spinner ml-3 inline' />}
                </button>
                <button
                  type='button'
                  className='cancel-button'
                  disabled={mutating || reverting}
                  onClick={cancelForm}
                >
                  {format('app.cancel')}
                  {reverting && <FaSpinner className='spinner ml-3 inline' />}
                </button>
              </div>
            </div>
          </form>
        </div>
      </div>
    </div>
  ) : <Unauthorized />
}

export default ProductForm
