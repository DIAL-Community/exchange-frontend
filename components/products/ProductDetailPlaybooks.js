import { useCallback } from 'react'
import { useIntl } from 'react-intl'
import PlaybookCard from '../playbooks/PlaybookCard'
import EditableSection from '../shared/EditableSection'

const ProductDetailPlaybooks = ({ product }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id) => formatMessage({ id }), [formatMessage])

  const displayModeBody = (
    <div className='grid grid-cols-1'>
      {product.playbooks.map(
        (playbook, playbookIdx) =>
          <PlaybookCard key={playbookIdx} playbook={playbook} listType='list' filterDisplayed newTab/>)
      }
    </div>
  )

  return (
    <EditableSection
      sectionHeader={format('playbook.header')}
      displayModeBody={displayModeBody}
    />
  )
}

export default ProductDetailPlaybooks
