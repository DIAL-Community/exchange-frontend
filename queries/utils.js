const MIN_SEARCH_VALUE_CHARACTERS = 2

export const fetchSelectOptions = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}

export const fetchSelectOptionsWithMature = async (client, input, query, fetchedDataCallback, locale = null) => {
  const variables = {
    search: input,
    mature: true
  }

  if (input && input.trim().length < MIN_SEARCH_VALUE_CHARACTERS) {
    return []
  } else if (locale) {
    variables.locale = locale
  }

  const response = await client.query({ query, variables })

  return response.data ? fetchedDataCallback(response.data) : []
}
