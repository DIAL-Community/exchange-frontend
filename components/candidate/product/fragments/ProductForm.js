import React, { useCallback, useContext, useRef, useState } from 'react'
import { useRouter } from 'next/router'
import ReCAPTCHA from 'react-google-recaptcha'
import { Controller, useFieldArray, useForm } from 'react-hook-form'
import { FaSpinner, FaXmark } from 'react-icons/fa6'
import { useIntl } from 'react-intl'
import { useMutation, useQuery } from '@apollo/client'
import { GRAPH_QUERY_CONTEXT } from '../../../../lib/apolloClient'
import { ToastContext } from '../../../../lib/ToastContext'
import { HtmlEditor } from '../../../shared/form/HtmlEditor'
import Input from '../../../shared/form/Input'
import Select from '../../../shared/form/Select'
import UrlInput from '../../../shared/form/UrlInput'
import ValidationError from '../../../shared/form/ValidationError'
import { handleLoadingQuery, handleQueryError } from '../../../shared/GraphQueryHandler'
import { CREATE_CANDIDATE_PRODUCT } from '../../../shared/mutation/candidateProduct'
import {
  CANDIDATE_PRODUCT_EXTRA_ATTRIBUTES_QUERY, CANDIDATE_PRODUCT_PAGINATION_ATTRIBUTES_QUERY,
  PAGINATED_CANDIDATE_PRODUCTS_QUERY
} from '../../../shared/query/candidateProduct'
import { DEFAULT_PAGE_SIZE } from '../../../utils/constants'

const MaintainerCompositeAttribute = ({ errors, control, register, extraAttribute }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const { title, required, attributes } = extraAttribute
  const { fields, append, remove } = useFieldArray({
    control,
    // TODO: Doesn't support dynamic name. https://www.react-hook-form.com/api/usefieldarray/
    // If we need more composite type, we need to copy and paste this section of the code.
    name: 'maintainers',
    rules: {
      required: required ? format('validation.required') : false
    }
  })

  return (
    <div className='flex flex-col gap-y-6' key={name}>
      <div className='text-sm font-medium'>
        {title}
      </div>
      {fields.map((item, index) => (
        <div className='flex flex-col gap-y-6 relative' key={item.id}>
          {attributes.map(({ name, title, required, description }) => (
            <div className='flex flex-col gap-y-2' key={`${index}-${name}`}>
              <label
                className={required ? 'required-field' : ''}
                htmlFor={`maintainers.${index}.${name}`}
              >
                {title}
              </label>
              <Input
                id={`maintainers.${index}.${name}`}
                {...register(`maintainers.${index}.${name}`, {
                  required: {
                    value: required,
                    message: format('validation.required')
                  }
                })}
                placeholder={description}
                isInvalid={errors[`maintainers.${index}.${name}`]}
              />
              {errors?.maintainers?.[index][name] &&
                <ValidationError value={errors?.maintainers?.[index][name]?.message} />
              }
            </div>
          ))}
          <div className='absolute top-0 right-0'>
            <button
              type="button"
              className='bg-red-500 text-white px-1 py-1 rounded-md ml-auto'
              onClick={() => remove(index)}
            >
              <FaXmark />
            </button>
          </div>
          <div className='border-t border-dashed' />
        </div>
      ))}
      <div className='flex'>
        <button
          type="button"
          className='bg-dial-meadow text-white px-4 py-2 rounded-md ml-auto'
          onClick={() => append({ maintainerName: '', maintainerEmail: '' })}
        >
          Append
        </button>
      </div>
      {errors.maintainers?.root &&
        <ValidationError value={errors.maintainers?.root?.message} />
      }
    </div>
  )
}

const ProductForm = React.memo(({ product }) => {
  const { formatMessage } = useIntl()
  const format = useCallback((id, values) => formatMessage({ id }, values), [formatMessage])

  const slug = product?.slug ?? ''

  const captchaRef = useRef()
  const [captchaValue, setCaptchaValue] = useState()

  const [mutating, setMutating] = useState(false)
  const [reverting, setReverting] = useState(false)

  const resolveDefaultValueByFieldKey = (fieldKey) => {
    let defaultValue
    if (product?.extraAttributes) {
      const extraAttribute = product.extraAttributes.find(attribute => attribute.name === fieldKey)
      defaultValue = extraAttribute?.value
    }

    return defaultValue
  }

  const {
    handleSubmit,
    register,
    control,
    formState: { errors }
  } = useForm({
    mode: 'onSubmit',
    reValidateMode: 'onChange',
    shouldUnregister: true,
    defaultValues: {
      name: product?.name ?? '',
      website: product?.website ?? '',
      repository: product?.repository ?? '',
      description: product?.description ?? '',
      submitterEmail: product?.submitterEmail ?? '',
      maintainers: resolveDefaultValueByFieldKey('maintainers') ?? []
    }
  })

  const { loading, data, error } = useQuery(CANDIDATE_PRODUCT_EXTRA_ATTRIBUTES_QUERY, {
    context: {
      headers: {
        ...(product ? GRAPH_QUERY_CONTEXT.EDITING : GRAPH_QUERY_CONTEXT.CREATING)
      }
    }
  })

  const { showSuccessMessage, showFailureMessage } = useContext(ToastContext)

  const router = useRouter()
  const { locale } = router

  const [updateProduct, { reset }] = useMutation(CREATE_CANDIDATE_PRODUCT, {
    refetchQueries: [{
      query: CANDIDATE_PRODUCT_PAGINATION_ATTRIBUTES_QUERY,
      variables: { search: '' },
      context: {
        headers: {
          ...GRAPH_QUERY_CONTEXT.VIEWING
        }
      }
    }, {
      query: PAGINATED_CANDIDATE_PRODUCTS_QUERY,
      variables: { search: '', limit: DEFAULT_PAGE_SIZE, offset: 0 },
      context: {
        headers: {
          ...GRAPH_QUERY_CONTEXT.VIEWING
        }
      }
    }],
    onCompleted: (data) => {
      const { createCandidateProduct: response } = data
      if (response.candidateProduct && response.errors.length === 0) {
        setMutating(false)
        const redirectPath = `/${locale}` +
          `/candidate/products/${response.candidateProduct.slug}`
        const redirectHandler = () => router.push(redirectPath)
        showSuccessMessage(
          format('toast.submit.success', { entity: format('ui.candidateProduct.label') }),
          redirectHandler
        )
      } else {
        showFailureMessage(format('toast.submit.failure', { entity: format('ui.candidateProduct.label') }))
        setMutating(false)
        reset()
      }
    },
    onError: () => {
      showFailureMessage(format('toast.submit.failure', { entity: format('ui.candidateProduct.label') }))
      setMutating(false)
      reset()
    }
  })

  const buildExtraAttributes = (otherFormValues) => {
    const extraAttributes = []
    if (data.candidateProductExtraAttributes) {
      const attributeDefinitions = data.candidateProductExtraAttributes

      return attributeDefinitions.map(attributeDefinition => {
        const extraAttribute = {}
        extraAttribute['name'] = attributeDefinition.name
        extraAttribute['value'] = otherFormValues[attributeDefinition.name]
        extraAttribute['type'] = attributeDefinition.type
        extraAttribute['index'] = attributeDefinition.index
        extraAttribute['title'] = attributeDefinition.title
        extraAttribute['description'] = attributeDefinition.description

        return extraAttribute
      })
    }

    return extraAttributes
  }

  const doUpsert = async (data) => {
    // Set the loading indicator.
    setMutating(true)
    // Pull all needed data from session and form.
    const {
      name,
      website,
      repository,
      description,
      submitterEmail,
      ...otherValues
    } = data

    // Send graph query to the backend. Set the base variables needed to perform update.
    const variables = {
      name,
      slug,
      website,
      repository,
      description,
      submitterEmail,
      extraAttributes: buildExtraAttributes(otherValues),
      captcha: captchaValue
    }

    updateProduct({
      variables,
      context: {
        headers: {
          'Accept-Language': locale
        }
      }
    })
  }

  const updateCaptchaData = (value) => {
    setCaptchaValue(value)
  }

  const cancelForm = () => {
    if (product) {
      setReverting(true)
      router.push(`/${locale}/candidate/products/${slug}`)
    }
  }

  // Handle useQuery return values.
  if (loading) {
    return handleLoadingQuery()
  } else if (error) {
    return handleQueryError(error)
  }

  // Parse the extra attributes information.
  const { candidateProductExtraAttributes } = data

  return (
    <form onSubmit={handleSubmit(doUpsert)}>
      <div className='px-4 lg:px-0 py-4 lg:py-6 text-dial-meadow'>
        <div className='flex flex-col gap-y-6 text-sm'>
          <div className='text-xl font-semibold'>
            {product
              ? format('app.editEntity', { entity: product.name })
              : `${format('app.createNew')} ${format('ui.product.label')}`}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='name'>
              {format('ui.candidateProduct.name')}
            </label>
            <Input
              {...register('name', { required: format('validation.required') })}
              id='name'
              placeholder={format('ui.candidateProduct.name.placeholder')}
              isInvalid={errors.name}
            />
            {errors.name && <ValidationError value={errors.name?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='website'>
              {format('ui.candidateProduct.website')}
            </label>
            <Controller
              id='website'
              name='website'
              control={control}
              render={({ field: { value, onChange } }) => (
                <UrlInput
                  id='website'
                  value={value}
                  onChange={onChange}
                  placeholder={format('ui.candidateProduct.website.placeholder')}
                />
              )}
            />
          </div>
          <div className='flex flex-col gap-y-2'>
            <label htmlFor='repository'>
              {format('ui.candidateProduct.repository')}
            </label>
            <Controller
              id='repository'
              name='repository'
              control={control}
              render={({ field: { value, onChange } }) => (
                <UrlInput
                  id='repository'
                  value={value}
                  onChange={onChange}
                  placeholder={format('ui.candidateProduct.repository.placeholder')}
                />
              )}
            />
          </div>
          <div className='flex flex-col gap-y-2'>
            <label id='description' className='required-field'>
              {format('ui.candidateProduct.description')}
            </label>
            <Controller
              name='description'
              control={control}
              render={({ field: { value, onChange } }) => (
                <HtmlEditor
                  labelledBy='description'
                  onChange={onChange}
                  initialContent={value}
                  placeholder={format('ui.candidateProduct.description.placeholder')}
                />
              )}
              rules={{ required: format('validation.required') }}
            />
            {errors.description && <ValidationError value={errors.description?.message} />}
          </div>
          <div className='flex flex-col gap-y-2'>
            <label className='required-field' htmlFor='submitterEmail'>
              {format('ui.candidateProduct.email')}
            </label>
            <Input
              {...register('submitterEmail', { required: format('validation.required') })}
              id='submitterEmail'
              placeholder={format('ui.candidateProduct.email.placeholder')}
              isInvalid={errors.submitterEmail}
            />
            {errors.submitterEmail && <ValidationError value={errors.submitterEmail?.message} />}
          </div>
          <div className='border-t border-dashed' />
          <div className='text-base font-semibold'>
            {format('ui.candidateProduct.extraAttributes')}
          </div>
          {candidateProductExtraAttributes.map((extraAttribute) => {
            const { name, title, description, type, required, multiple, options } = extraAttribute
            if (type === 'composite') {
              return (
                <MaintainerCompositeAttribute
                  key={name}
                  errors={errors}
                  control={control}
                  register={register}
                  extraAttribute={extraAttribute}
                />
              )
            } else if (type === 'text') {
              return (
                <div className='flex flex-col gap-y-2' key={name}>
                  <label className={required ? 'required-field' : ''} htmlFor={name}>
                    {title}
                  </label>
                  <Input
                    id={name}
                    {...register(name, {
                      required: {
                        value: required,
                        message: format('validation.required')
                      }
                    })}
                    defaultValue={resolveDefaultValueByFieldKey(name)}
                    placeholder={description}
                    isInvalid={errors[name]}
                  />
                  {errors[name] && <ValidationError value={errors[name]?.message} />}
                </div>
              )
            } else if (type === 'url') {
              return (
                <div className='flex flex-col gap-y-2' key={name}>
                  <label className={required ? 'required-field' : ''} htmlFor={name}>
                    {title}
                  </label>
                  <Controller
                    id={name}
                    name={name}
                    control={control}
                    defaultValue={resolveDefaultValueByFieldKey(name) ?? ''}
                    render={({ field: { value, onChange } }) => (
                      <UrlInput
                        id={name}
                        value={value}
                        onChange={onChange}
                        placeholder={description}
                      />
                    )}
                    rules={{
                      required: {
                        value: required,
                        message: format('validation.required')
                      }
                    }}
                  />
                  {errors[name] && <ValidationError value={errors[name]?.message} />}
                </div>
              )
            } else if (type === 'select') {
              return (
                <div className='flex flex-col gap-y-2' key={name}>
                  <label className={required ? 'required-field' : ''} htmlFor={name}>
                    {title}
                  </label>
                  <Controller
                    id={name}
                    name={name}
                    control={control}
                    defaultValue={resolveDefaultValueByFieldKey(name)}
                    rules={{
                      required: {
                        value: required,
                        message: format('validation.required')
                      }
                    }}
                    render={({ field: { value, onChange } }) => (
                      <Select
                        id={name}
                        value={value}
                        onChange={onChange}
                        placeholder={description}
                        options={options.map((option) => ({
                          label: option,
                          value: option
                        }))}
                        isMulti={multiple}
                      />
                    )}
                  />
                  {errors[name] && <ValidationError value={errors[name]?.message} />}
                </div>
              )
            }
          })}
          <div className='border-t border-dashed' />
          <ReCAPTCHA
            sitekey={process.env.NEXT_PUBLIC_CAPTCHA_SITE_KEY}
            onChange={updateCaptchaData}
            ref={captchaRef}
          />
          <div className='flex flex-wrap text-base mt-6 gap-3'>
            <button
              type='submit'
              className='submit-button'
              disabled={mutating || reverting || !captchaValue}
            >
              {`${format('app.submit')} ${format('ui.product.label')}`}
              {mutating && <FaSpinner className='spinner ml-3' />}
            </button>
            {product &&
              <button type='button'
                className='cancel-button'
                disabled={mutating || reverting}
                onClick={cancelForm}
              >
                {format('app.cancel')}
                {reverting && <FaSpinner className='spinner ml-3' />}
              </button>
            }
          </div>
        </div>
      </div>
    </form>
  )
})

ProductForm.displayName = 'ProductForm'

export default ProductForm
