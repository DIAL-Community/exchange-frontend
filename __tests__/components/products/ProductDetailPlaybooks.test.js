import ProductDetailPlaybooks from '../../../components/products/ProductDetailPlaybooks'
import { render } from '../../test-utils'
import { mockNextUseRouter } from '../../utils/nextMockImplementation'
import { product } from './data/ProductForm'

mockNextUseRouter()
describe('Unit tests for the ProductDetailPlaybooks component.', () => {
  test('Should match snapshot', () => {
    const { container } = render(
      <ProductDetailPlaybooks
        product={product}
      />
    )
    expect(container).toMatchSnapshot()
  })
})
