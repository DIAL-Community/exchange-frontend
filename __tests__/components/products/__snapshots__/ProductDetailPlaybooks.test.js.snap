// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Unit tests for the ProductDetailPlaybooks component. Should match snapshot 1`] = `
<div>
  <div
    class="Toastify"
  />
  <div
    class="mt-12"
    data-testid="editable-section"
  >
    <div
      class="flex justify-between items-center mb-3"
    >
      <div
        class="card-title text-dial-sapphire"
      >
        Playbooks
      </div>
      <div
        class="gap-3"
      />
    </div>
    <div
      class="grid grid-cols-1"
    >
      <a
        href="/playbooks/playbook_1"
        target="_blank"
      >
        <div
          class="cursor-pointer hover:rounded-lg hover:shadow-lg border-3 border-transparent hover:border-dial-sunshine"
        >
          <div
            class="bg-white border border-dial-gray shadow-lg rounded-md"
          >
            <div
              class="relative flex flex-row gap-x-2 lg:gap-x-4 px-4 py-6"
            >
              <div
                class="w-10/12 lg:w-6/12 flex gap-3 text-dial-gray-dark my-auto"
              >
                <div
                  class="block w-8 relative"
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
                  >
                    <img
                      alt="Logo for: Playbook 1"
                      data-nimg="fill"
                      data-tip="Playbook: Playbook 1"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: scale-down; object-position: left;"
                    />
                    <noscript />
                  </span>
                </div>
                <div
                  class="ml-2 mt-0.5 w-full h-3/5 font-semibold line-clamp-1"
                >
                  Playbook 1
                </div>
              </div>
              <div
                class="flex gap-1.5 text-sm lg:ml-auto"
              >
                <div
                  class="flex flex-row"
                  data-testid="horizontal-items-list"
                >
                  <div
                    class="flex flex-row flex-wrap overflow-hidden max-h-8 gap-1"
                  >
                    <div
                      class="bg-dial-alice-blue p-2 rounded"
                      data-testid="items-not-applicable"
                    >
                      N/A
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </a>
      <a
        href="/playbooks/playbook_2"
        target="_blank"
      >
        <div
          class="cursor-pointer hover:rounded-lg hover:shadow-lg border-3 border-transparent hover:border-dial-sunshine"
        >
          <div
            class="bg-white border border-dial-gray shadow-lg rounded-md"
          >
            <div
              class="relative flex flex-row gap-x-2 lg:gap-x-4 px-4 py-6"
            >
              <div
                class="w-10/12 lg:w-6/12 flex gap-3 text-dial-gray-dark my-auto"
              >
                <div
                  class="block w-8 relative"
                >
                  <span
                    style="box-sizing: border-box; display: block; overflow: hidden; background: none; opacity: 1; border: 0px; margin: 0px; padding: 0px; position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px;"
                  >
                    <img
                      alt="Logo for: Playbook 2"
                      data-nimg="fill"
                      data-tip="Playbook: Playbook 2"
                      decoding="async"
                      src="data:image/gif;base64,R0lGODlhAQABAIAAAAAAAP///yH5BAEAAAAALAAAAAABAAEAAAIBRAA7"
                      style="position: absolute; top: 0px; left: 0px; bottom: 0px; right: 0px; box-sizing: border-box; padding: 0px; margin: auto; display: block; width: 0px; height: 0px; min-width: 100%; max-width: 100%; min-height: 100%; max-height: 100%; object-fit: scale-down; object-position: left;"
                    />
                    <noscript />
                  </span>
                </div>
                <div
                  class="ml-2 mt-0.5 w-full h-3/5 font-semibold line-clamp-1"
                >
                  Playbook 2
                </div>
              </div>
              <div
                class="flex gap-1.5 text-sm lg:ml-auto"
              >
                <div
                  class="flex flex-row"
                  data-testid="horizontal-items-list"
                >
                  <div
                    class="flex flex-row flex-wrap overflow-hidden max-h-8 gap-1"
                  >
                    <div
                      class="bg-dial-alice-blue p-2 rounded"
                      data-testid="items-not-applicable"
                    >
                      N/A
                    </div>
                  </div>
                </div>
              </div>
            </div>
          </div>
        </div>
      </a>
    </div>
  </div>
</div>
`;
